File: Data Structures & Algorithms/sum-of-two-integers/submission-1.cpp
class Solution {
public:
    int getSum(int a, int b) {
        while (b != 0) {
            int sum = a^b, carry = (a&b)<<1;
            a = sum;
            b = carry;
        }

        return a;
    }
};

/*
    a = 4, b = 7
    a = 100, b = 111
    s = 100^111 = 011, c = 100&111<<1 = 1000, a = s, b = c

    a = 11, b = 1000
    s = 0011^1000 = 1011, c = 0
*/

File: Data Structures & Algorithms/climbing-stairs/submission-1.java
class Solution {
    public int climbStairs(int n) {
        if (n <= 1) {
            return 1;
        }

        int[] dp = new int[n+1];
        int x = 1, y = 1;
        int z = 1;
    
        for (int i = 2; i <= n; i++) {
            z = x + y;
            y = x;
            x = z;
        }

        return z;
    }
}
